common/include/pcl/pcl_macros.h

```diff
@@ -67,24 +67,53 @@
 
 #include <pcl/pcl_config.h>
 
+#include <boost/preprocessor/arithmetic/sub.hpp>
+#include <boost/preprocessor/comparison/equal.hpp>
+#include <boost/preprocessor/comparison/greater.hpp>
+#include <boost/preprocessor/control/if.hpp>
+
 // It seems that __has_cpp_attribute doesn't work correctly
 // when compiling with some versions of nvcc so we
 // additionally check if nvcc is used before setting the
-// PCL_DEPRECATED macro to [[deprecated]].
+// PCL_DEPRECATED_IMPL macro to [[deprecated]].
 #if defined(__has_cpp_attribute) && __has_cpp_attribute(deprecated) && !defined(__CUDACC__)
-  #define PCL_DEPRECATED(message) [[deprecated(message)]]
+  #define PCL_DEPRECATED_IMPL(message) [[deprecated(message)]]
 #elif defined(__GNUC__) || defined(__clang__)
-  #define PCL_DEPRECATED(message) __attribute__((deprecated(message)))
+  #define PCL_DEPRECATED_IMPL(message) __attribute__((deprecated(message)))
 #elif defined(_MSC_VER)
   // Until Visual Studio 2013 you had to use __declspec(deprecated).
   // However, we decided to ignore the deprecation for these version because
   // of simplicity reasons. See PR #3634 for the details.
-  #define PCL_DEPRECATED(message)
+  #define PCL_DEPRECATED_IMPL(message)
 #else
-  #warning "You need to implement PCL_DEPRECATED for this compiler"
-  #define PCL_DEPRECATED(message)
+  #warning "You need to implement PCL_DEPRECATED_IMPL for this compiler"
+  #define PCL_DEPRECATED_IMPL(message)
 #endif
 
+/**
+ * \brief PCL removes deprecations after 2 release cycles on minor version
+ */
+#define PCL_DEPRECATION_CYCLE 2
+
+#define DEPRECATED_MINOR(minor, msg)                                                      \
+    BOOST_PP_IF(BOOST_PP_GREATER(BOOST_PP_SUB(PCL_MINOR_VERSION, PCL_DEPRECATION_CYCLE)), \
+                                 minor                                                    \
+                PCL_DEPRECATED_IMPL(msg),                                                 \
+                unneeded_deprecation)
+
+/**
+ * \brief macro for compatibility across compilers and help remove old deprecated
+ *        items before release based on PCL_DEPRECATION_CYCLE
+ *
+ * \detail compiler errors of `unneeded_deprecation` and `major_version_mismatch`
+ * are hints to the developer that those items ca be safely removed
+ */
+#define PCL_DEPRECATED(major, minor, message) \
```

Comment:
`major` and `minor` are what exactly? Last version where the item was not yet deprecated? First version where it was deprecated? I'm not sure how we will choose this version given that we don't always know what the next released version will end up being (patch or minor).

I also feel like `message` is more important should be the first argument.

Finally, we may consider appending (computed) target removal version to the message.

Reply:
> major and minor are what exactly?

First version where it was deprecated, as per implementation. I can switch it to mean till when it'd be supported. Then I can convert the message to: "slated for removal in vX.Y. <usr msg>"

Reply:
Hm, I think I like this. No need to do arithmetics in macro and `PCL_DEPRECATION_CYCLE`.

Reply:
PTAL